main.c

```diff
@@ -69,21 +69,61 @@ static const char MAIN_ISR_DOSID_String[] = "Crazii  SBEMU   Sound Blaster emula
 static void MAIN_TSR_InstallationCheck();
 static void MAIN_TSR_Interrupt();
 
+uint16_t main_hw_fmport = 0;
+#define hw_fm_outb(reg,data) outp(main_hw_fmport+reg,data)
+#define hw_fm_inb(reg) inp(main_hw_fmport+reg)
+
+uint16_t main_hw_mpuport = 0;
+#define hw_mpu_outb(reg,data) outp(main_hw_mpuport+reg,data)
+#define hw_mpu_inb(reg) inp(main_hw_mpuport+reg)
+
 static uint32_t MAIN_OPL3_388(uint32_t port, uint32_t val, uint32_t out)
 {
-    return out ? OPL3EMU_PrimaryWriteIndex(val) : OPL3EMU_PrimaryRead(val);
+  if (main_hw_fmport) {
```

Comment:
Is it a problem for some slow machines if you check `main_hw_fmport` every time?
I think it is OK to commit it as is now, and improve it later.

There needs to be a way to have two or more cards selected. For example YMF for FM/MIDI and NULL or HDA for sound effects.
You could have options like /SC for sound effects, /FC for FM, /MC for MIDI.

Reply:
I see. Naturally I think you would add some routines card_fm_read, card_fm_write etc. to struct one_sndcard_info.
I can do this later, but I want to clean up the Linux drivers first.
It would be easier for me to merge to/from my fork if this YMF patch gets merged first.

Before this I didn't have any experience with DOS.
I thought since this is a TSR it needs to be fast/simple, and, looking at main.c, thought that a global variable (`main_hw_fmport`) was OK.

Reply:
OK, we can leave it for later. I'm really sorry for keep you waiting so long, I was digging into the doom crash problem and now still be. :)

Reply:
what is the problem with doom? With which card?

Reply:
> what is the problem with doom? With which card?

The onboard HDA that @hjnijlunsing has, it's a compatibility problem not quite related to the soundcard driver as opl3 works.